Generate process metrics with `semconv` yaml

braydonk opened the pull request:
Supplants #160 

## Changes

This PR adds a yaml version of the process metrics and changes the process metrics documentation to be generated from the yaml instead of hand-written.

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] [CHANGELOG.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CHANGELOG.md) updated for non-trivial changes.
* [x] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

mx-psi commented:
Needs minor edits for CI to pass but content LGTM

braydonk commented:
> Needs minor edits for CI to pass

Thanks @mx-psi those are fixed now, also added something to `.yamllint` to not scan `node_modulues` when running `yamllint` locally (didn't think it was worth a whole separate PR for something that small).

braydonk commented on docs/system/process-metrics.md:
```diff
@@ -20,32 +20,179 @@ metrics](runtime-environment-metrics.md).
+<!-- semconv metric.process.threads(metric_table) -->
+| Name     | Instrument Type | Unit (UCUM) | Description    |
+| -------- | --------------- | ----------- | -------------- |
+| `process.threads` | UpDownCounter | `{thread}` | Process threads count. |
```
Hmm I'm not sure, it looks like with these `process.thread` fields in the linked docs, they all seem to be associated with a single thread itself as a resource, whereas the thread count is associated with a process resource. I worry that might be difficult to understand at first.

Regardless I will include it in the following issue for things to revisit for these metrics (kind of like #325). My plan is to open it once this is merged.

AlexanderWert replied:
See this related issue / discussion: https://github.com/open-telemetry/semantic-conventions/issues/260

joaopgrassi replied:
@braydonk could you also please add an issue for this, and add to the stability system project/label `system-semconv-wg-ga-blocker`

braydonk replied:
I ended up changing it to `process.thread.count` in the end, so I don't think we need an issue for it.

joaopgrassi replied:
Ah right. Going to resolve this discussion then if it's fine for everyone.

braydonk commented:
Looks like these workflows require approval now if someone can press the button for me. :smile:

ChrsMark reviewed (approved):
Content-wise lgtm! 

@braydonk please file an issue for https://github.com/open-telemetry/semantic-conventions/pull/330#discussion_r1330143868 so as to examine https://github.com/open-telemetry/semantic-conventions/issues/260 as well :).

braydonk commented:
Thanks, I'll file an issue today and I'll put my thoughts on #260 in there!

joaopgrassi commented:
I will try to take a look at it today or tomorrow!

joaopgrassi commented on CHANGELOG.md:
```diff
@@ -7,6 +7,20 @@ release.
 
+- BREAKING: Generate process metrics from YAML
+  ([#330](https://github.com/open-telemetry/semantic-conventions/pull/330))
+  - Rename attributes for `process.cpu.time`
```
If you want, this can be shorter by doing the same as I did for the system metrics:

```
- Rename attributes for `system.cpu.*` metrics:
    - `state` to `system.cpu.state`
    - `cpu` to `system.cpu.logical_number`
```

braydonk replied:
Resolved in [5ad3a95](https://github.com/open-telemetry/semantic-conventions/pull/330/commits/5ad3a95ab2ddaf681fef56fbe757d1dedda68744) (rebased so it moved to a different place)

joaopgrassi commented on model/metrics/process-metrics.yaml:
```diff
@@ -0,0 +1,136 @@
+    type: metric
+    metric_name: process.paging.faults
+    brief: "Number of page faults the process has made."
+    instrument: updowncounter
```
In the current conventions, this is a counter.

braydonk replied:
Resolved in [5ad3a95](https://github.com/open-telemetry/semantic-conventions/pull/330/commits/5ad3a95ab2ddaf681fef56fbe757d1dedda68744)

joaopgrassi commented on schema-next.yaml:
```diff
@@ -131,6 +131,24 @@ versions:
             apply_to_metrics:
               - system.processes.count
+        # https://github.com/open-telemetry/semantic-conventions/pull/330
+        - rename_attributes:
```
I think it's better to read if the attributes listed here are split by their "namespace". Like I did for the system metrics https://github.com/open-telemetry/semantic-conventions/pull/89/files#diff-3ddd62d7281d98d63c3e92f5ae8cd7776971b44ca39f52b0140dc02363dfebda

braydonk replied:
Resolved in [5ad3a95](https://github.com/open-telemetry/semantic-conventions/pull/330/commits/5ad3a95ab2ddaf681fef56fbe757d1dedda68744) (rebased so it moved to a different place)

joaopgrassi commented on docs/system/process-metrics.md:
```diff
@@ -20,32 +20,179 @@ metrics](runtime-environment-metrics.md).
+
+<!-- semconv metric.process.threads(full) -->
+<!-- endsemconv -->
+
```
We are missing `process.open_file_descriptors`

braydonk replied:
Resolved in [5ad3a95](https://github.com/open-telemetry/semantic-conventions/pull/330/commits/5ad3a95ab2ddaf681fef56fbe757d1dedda68744)

joaopgrassi reviewed (commented):
Here's a first pass

joaopgrassi commented on model/metrics/process-metrics.yaml:
```diff
@@ -0,0 +1,136 @@
+    instrument: counter
+    unit: "By"
+    attributes:
+      - id: process.disk.io.direction
```
The attribute must be under the namespace (`process.disk.*`) here as well

braydonk commented:
It's possible I'm missing some kind of context from discussions around #51, but it seems like what I have here should be fine. With the paging fault type attribute for example: it is in the `process.paging.faults` namespace. I added that namespace, as the attribute was previously `type` (just `type`). `process.paging` is not any more of a proper namespace than `process.paging.faults` as far as I can tell.

If I'm indeed missing something extra that's not from #51, and this attribute should go to `process.paging`, then I can also rename it to `fault_type`. But I would like to understand why for this and the other pointed out attributes are not following the decision from #51 already.

joaopgrassi commented:
> With the paging fault type attribute for example: it is in the `process.paging.faults` namespace

The thing I'm not sure if that `process.paging.faults` is the a metric, not a namespace. From reading #51 and the linked issues, my interpretation is that attributes should go on the namespace , meaning `process.paging.*` in this case. Maybe we need clarification if attributes can also be nested under the "implicit metric namespace". 

CC @open-telemetry/specs-semconv-approvers @open-telemetry/specs-semconv-maintainers

joaopgrassi commented on model/metrics/process-metrics.yaml:
```diff
@@ -0,0 +1,136 @@
+    instrument: counter
+    unit: "{fault}"
+    attributes:
+      - id: process.paging.faults.type
```
I opened an issue to discuss it #394. I noticed the decision is also not documented anywhere and we also have the other discussion about "metric names should be allowed to be namespaces" https://github.com/open-telemetry/semantic-conventions/issues/50

ChrsMark replied:
Maybe going with sth like `process.paging.fault.count` and `process.paging.fault.type` would be more tidy?
This would also be aligned with ECS' alignment effort and not having fields conflicting with leaf nodes as described at https://github.com/open-telemetry/semantic-conventions/pull/333#issuecomment-1764384113.

braydonk commented:
I have updated the PR to match the semantic convention guidance as well as I'm able.

I renamed the metrics that are `updowncounter`s to use the `.count` pluralization method. I also changed the metric attributes such that they do not use the metric names as namespaces.

braydonk commented on model/metrics/process-metrics.yaml:
```diff
@@ -0,0 +1,136 @@
+    type: attribute_group
+    brief: "Attributes for process CPU metrics."
+    attributes:
+      - id: state
```
For what it's worth, I highly doubt these attributes will ever be used outside of this namespace. Is it still common practice to put these in the global registry even if it only makes sense for the attribute to belong to a single namespace?

ChrsMark replied:
I guess for consistency reasons all attributes should go to the registry no matter how many refs exist for them.

ChrsMark replied:
I'm fine with doing so. In that case let's file an issue to keep track of this unless the follow up PR comes right after :).

joaopgrassi replied:
@braydonk created the issue here #650

joaopgrassi commented on docs/system/process-metrics.md:
```diff
@@ -20,32 +20,191 @@ metrics](/docs/runtime/README.md#metrics).
 - [Metric Instruments](#metric-instruments)
-  * [Process](#process)
-- [Attributes](#attributes)
+  * [Metric: `process.cpu.time`](#metric-processcputime)
```
Maybe consider making the section follow the same structure as in the system metrics markdown: https://github.com/braydonk/semantic-conventions/blob/process_metrics/docs/system/system-metrics.md?plain=1#L24

joaopgrassi commented on docs/system/process-metrics.md:
```diff
@@ -20,32 +20,191 @@ metrics](/docs/runtime/README.md#metrics).
 ## Metric Instruments
 
-### Process
+### Metric: `process.cpu.time`
```
We also specify here if the metric is required/recommended https://github.com/braydonk/semantic-conventions/blob/process_metrics/docs/system/system-metrics.md?plain=1#L68

braydonk replied:
I think every metric in this document is going to be recommended. Should I add this to every section, or include something in the top?

joaopgrassi replied:
I'd say yes, to be consistent with the other files. I have opened an issue in the past, to make this part of the markdown template in the built-tool repo.

braydonk replied:
Done in [914e018](https://github.com/open-telemetry/semantic-conventions/pull/330/commits/914e018823e3bcb85fc91eddb9b948ed4d13c2ea)

joaopgrassi reviewed (commented):
I think this looks good now. Just a few markdown structural nits

braydonk commented:
I added `proposed_changelog.md` and `proposed_schema_next.yaml`, which upon PR approval I will add properly to the `CHANGELOG.md` and `schema-next.yaml` files. This is to avoid constantly needing to rebase these giant files if the PR lasts over a long period of time.

joaopgrassi reviewed (approved):
LGTM after fixing the CI issues

ChrsMark reviewed (approved):
LGTM. Let's make sure that we file an issue for moving the attributes to the registry: https://github.com/open-telemetry/semantic-conventions/pull/330#discussion_r1419228700

braydonk commented:
> LGTM. Let's make sure that we file an issue for moving the attributes to the registry: https://github.com/open-telemetry/semantic-conventions/pull/330#discussion_r1419228700

Opened https://github.com/open-telemetry/semantic-conventions/issues/650 to track this.

braydonk commented:
This should be ready for merge.

joaopgrassi commented:
I will bring this up in the semconv SIG meeting today, to raise awareness. I guess we can merge it then later today/tomorrow if no objections.
